core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -133,6 +134,10 @@ public class RemoteLogManager implements Closeable {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(RemoteLogManager.class);
     private static final String REMOTE_LOG_READER_THREAD_NAME_PREFIX = "remote-log-reader";
+    private static final Set<RemoteLogSegmentState> SEGMENT_DELETION_VALID_STATES = Collections.unmodifiableSet(EnumSet.of(
```

Comment:
Great catch. I guess the transition hasn't considered the current implementation where copying and deletion are sequential. If the implementation changes, it may be possible to have this scenario.
I'd say it should be considered a valid transition. I will include it in the valid states and adapt the test case.

Reply:
COPY_SEGMENT_STARTED segments are eligible for deletion when those segments were not able to be copied by the leader as the leader went through ungraceful shutdown or for any oher reasons. New leader may pickup the resepctive segments for the targeted offsets that need to be copied and the earlier failed segment will remain in the COPY_SEGMENT_STARTED state and it will eventually be deleted by retention cleanup logic. 

So, COPY_SEGMENT_STARTED is a valid transition even now when copy and deletion are happening sequentially.

Reply:
@satishd in that case wouldn't the new leader retry to upload the segment? If a segment is left on copy_started state, then it'd risk some segment not being uploaded. Is this expected?

Reply:
@jeqo Leader will upload the required segment with a new UUID by finding the offset that needs to be copied as mentioned [here](https://github.com/apache/kafka/blob/trunk/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L618). It will not change the state of the failed copy of an earlier segment.

Reply:
Right, one way to check is whether there is a segment associated with the target offset or not by using [RLMM#remoteLogSegmentMetadata()](https://github.com/apache/kafka/blob/trunk/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManager.java#L123).